### PR TITLE
Change stdlib swiftc's to depend on .swiftmodule's instead of .o's

### PR DESCRIPTION
**Pull request by cooperp**
Currently SwiftOnoneSupport.o waits on Swift.o to finish compiling before we start
compiling SwiftOnoneSupport.o.  We don't need the .o for compilation but instead just
need Swift.swiftmodule which we were generating anyway.

This change causes .o and { .swiftmodule, .swiftdoc } to be built independently of each
other.  Downstream compiles only wait on the module while downstream links wait on the .o.

Given that .apinotes are sometimes used by swiftc invocations, this also separates out
apinotes in to its own step which these other commands can both depend on.

The result of this is a 30% improvement in building the stdlib with -R and 20% with -d.
Specifically, -R drops from 3m to 2m and -d drops from 12m to just under 10m.

This was tested with --validation-test which uncovered a missing dependency.  After adding that
in, we now pass the tests.

### CODE REVIEW

**Comment by DougGregor** [comment]
Fantastic!

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by rudkx** [comment]
Yay! Thanks Pete.

**Comment by DougGregor** [comment]
Would love a review from someone more familiar with our CMake setup.

File: cmake/modules/AddSwift.cmake [hunk 1]
```diff
@@ -1320,6 +1329,8 @@ function(add_swift_library name)
         # Configure variables for this subdirectory.
         set(VARIANT_SUFFIX "-${SWIFT_SDK_${sdk}_LIB_SUBDIR}-${arch}")
         set(VARIANT_NAME "${name}${VARIANT_SUFFIX}")
+        set(MODULE_VARIANT_SUFFIX "-swiftmodule-${SWIFT_SDK_${sdk}_LIB_SUBDIR}-${arch}")
```
Comment (jrose-apple) [inline cmake/modules/AddSwift.cmake]: Should it just get this from the regular VARIANT_SUFFIX?

File: cmake/modules/AddSwift.cmake [hunk 2]
```diff
@@ -1384,6 +1389,15 @@ function(add_swift_library name)
           endif()
         endforeach()
 
+        set(swiftlib_framework_depends_flattened ${SWIFTLIB_FRAMEWORK_DEPENDS})
```
Comment (jrose-apple) [inline cmake/modules/AddSwift.cmake]: Why move this?

File: cmake/modules/SwiftSource.cmake [hunk 3]
```diff
@@ -284,16 +291,16 @@ function(_compile_swift_files dependency_target_out_var_name)
...
       set(module_doc_file "${module_dir}/${SWIFTFILE_MODULE_NAME}.swiftdoc")
-      list(APPEND swift_flags
-          "-parse-as-library"
-          "-emit-module" "-emit-module-path" "${module_file}")
-    else()
-      list(APPEND swift_flags "-parse-as-library")
+      list(APPEND module_command
+          "-emit-module"
+          "-emit-module-path" "${module_file}"
+          "-o" "${module_file}")
```
Comment (jrose-apple) [inline cmake/modules/SwiftSource.cmake]: You shouldn't need both `-emit-module-path` and `-o` here. Either one should be fine.

File: cmake/modules/SwiftSource.cmake [hunk 4]
```diff
@@ -383,29 +390,66 @@ function(_compile_swift_files dependency_target_out_var_name)
...
+    set(command_touch_module_outputs
+      COMMAND "${CMAKE_COMMAND}" -E touch ${module_outputs})
+  endif()
+
+  # First generate the api notes if we need them.
+  list(LENGTH command_create_apinotes num_apinotes_outputs)
+  if (NOT ${num_apinotes_outputs} EQUAL 0)
+    add_custom_command_target(
+        api_notes_dependency_target
+        ${command_create_dirs}
```
Comment (jrose-apple) [inline cmake/modules/SwiftSource.cmake]: Creating directories needs to be done regardless, no?

File: cmake/modules/SwiftSource.cmake [hunk 5]
```diff
@@ -383,29 +390,66 @@ function(_compile_swift_files dependency_target_out_var_name)
...
+    set(command_touch_standard_outputs
+      COMMAND "${CMAKE_COMMAND}" -E touch ${standard_outputs})
+    set(command_touch_apinotes_outputs
+      COMMAND "${CMAKE_COMMAND}" -E touch ${apinotes_outputs})
+    set(command_touch_module_outputs
+      COMMAND "${CMAKE_COMMAND}" -E touch ${module_outputs})
+  endif()
+
+  # First generate the api notes if we need them.
+  list(LENGTH command_create_apinotes num_apinotes_outputs)
```
Comment (jrose-apple) [inline cmake/modules/SwiftSource.cmake]: Should this be measuring `apinotes_outputs` instead?

Is there any reason to not just use `if (apinotes_outputs)`?

File: cmake/modules/SwiftSource.cmake [hunk 6]
```diff
@@ -383,29 +390,66 @@ function(_compile_swift_files dependency_target_out_var_name)
...
+        # Create API notes before compiling, because this will affect the APIs
+        # the overlay sees.
+        ${command_create_apinotes}
+        ${command_touch_apinotes_outputs}
+        COMMAND ""
+        OUTPUT ${apinotes_outputs}
+        DEPENDS
+          ${swift_compiler_tool_dep}
+          ${source_files} ${SWIFTFILE_DEPENDS}
+          ${swift_ide_test_dependency} ${depends_create_apinotes}
```
Comment (jrose-apple) [inline cmake/modules/SwiftSource.cmake]: This should actually only depend on `depends_create_apinotes`. API notes are completely independent from any of the Swift stuff.

File: cmake/modules/SwiftSource.cmake [hunk 7]
```diff
@@ -383,29 +390,66 @@ function(_compile_swift_files dependency_target_out_var_name)
...
+        "${line_directive_tool}" "${source_files}" --
+        "${swift_compiler_tool}" "${module_command}" ${swift_flags}
+        "${source_files}"
+      ${command_touch_module_outputs}
+      OUTPUT ${module_outputs}
+      DEPENDS
+        ${swift_compiler_tool_dep}
+        ${source_files} ${SWIFTFILE_DEPENDS}
+        ${swift_ide_test_dependency} ${api_notes_dependency_target}
+      COMMENT "Compiling ${first_output}")
```
Comment (jrose-apple) [inline cmake/modules/SwiftSource.cmake]: Different comment, maybe?

**Review by jrose-apple — COMMENTED** [review]
It pretty much all looks fine to me; just a few comments here and there. I hadn't even thought to come at this from this side because I was still thinking of type-checking as the main slowdown. This does do some extra work and spawns more processes, but with an optimized stdlib it's probably a win. (Less sure about a debug compiler / debug stdlib combination, but that's not very common.)

Thanks, Pete!

File: cmake/modules/AddSwift.cmake [hunk 8]
```diff
@@ -536,7 +536,7 @@ endfunction()
 #
 # source1 ...
 #   Sources to add into this library
-function(_add_swift_library_single target name)
+function(_add_swift_library_single target module_target name)
```
Comment (gottesmm) [inline cmake/modules/AddSwift.cmake]: Can you add documentation to the header for what module_target is. Also, even better, can you change these options here to be cmake_parse_argument Arguments. Then below when you use "" for the module_target name, it would be at least self documenting that it is not a bug (see question below).

File: cmake/modules/AddSwift.cmake (hunk 2 above)
Comment (gottesmm) [inline cmake/modules/AddSwift.cmake]: And if it is just a move, can you do the move in a separate commit?

File: cmake/modules/AddSwift.cmake [hunk 9]
```diff
@@ -700,6 +701,13 @@ function(_add_swift_library_single target name)
       INSTALL_IN_COMPONENT "${SWIFTLIB_INSTALL_IN_COMPONENT}")
   add_swift_source_group("${SWIFTLIB_SINGLE_EXTERNAL_SOURCES}")
 
+  # If there were any swift sources, then a .swiftmodule may have been created.
+  # If that is the case, then add a target which is an alias of the module files.
+  if(NOT "${module_target}" STREQUAL "" AND NOT "${swift_module_dependency_target}" STREQUAL "")
```
Comment (gottesmm) [inline cmake/modules/AddSwift.cmake]: If you made module_target a cmake_parse_arguments target, then this would be cleaner. Specifically, you could do:

if (NOT "${SWIFTLIB_MODULE_TARGET}" AND NOT "${swift_module_dependency_target}")

endif()

Then it would be clear that module_target is an argument due to the SWIFTLIB_* prefix.

File: cmake/modules/AddSwift.cmake [hunk 10]
```diff
@@ -1352,23 +1363,17 @@ function(add_swift_library name)
               ${SWIFTLIB_SWIFT_MODULE_DEPENDS_LINUX})
         endif()
 
+        # Swift compiles depend on swift modules, while links depend on
+        # linked libraries.  Find targets for both of these here.
         set(swiftlib_module_dependency_targets)
+        set(swiftlib_private_link_libraries_targets)
```
Comment (gottesmm) [inline cmake/modules/AddSwift.cmake]: Unless I am missing something, isn't the variable swiftlib_private_link_libraries_targets dead?

File: cmake/modules/AddSwift.cmake [hunk 11]
```diff
@@ -1603,6 +1618,7 @@ function(add_swift_library name)
 
     _add_swift_library_single(
       ${name}
+      ""
```
Comment (gottesmm) [inline cmake/modules/AddSwift.cmake]: This is why I would suggest adding a named argument /assuming/ that is the correct thing to do here.

Then it would be clear that this is not a bug since you would have:

MODULE_TARGET ""

That being said, if you do what I suggested above, then this argument could be removed.

File: cmake/modules/SwiftSource.cmake [hunk 12]
```diff
@@ -274,6 +280,7 @@ function(_compile_swift_files dependency_target_out_var_name)
   set(module_file)
   set(module_doc_file)
 
+  set(module_command)
   if(NOT SWIFTFILE_IS_MAIN)
```
Comment (gottesmm) [inline cmake/modules/SwiftSource.cmake]: This section may make sense to refactor into its own function... thoughts?

Comment (gottesmm) [inline cmake/modules/SwiftSource.cmake]: My reasoning is that in the general flow of things, adding swift_flags -parse-as-library where you adding it, seems funky. That being said, this isn't that big of a deal.

**Review by gottesmm — CHANGES_REQUESTED** [review]
Meta request. Can you make sure that before/after if you do a clean build nothing is rebuilt. From what I understand ChrisB got us to that point and I would like to stay that way = ).

File: cmake/modules/AddSwift.cmake (hunk 1 above)
Comment (cooperp) [inline cmake/modules/AddSwift.cmake]: Do you mean the "${SWIFT_SDK_${sdk}_LIB_SUBDIR}-${arch}" part in MODULE_VARIANT_SUFFIX?  If so, then yeah, I'll fix that.

File: cmake/modules/AddSwift.cmake (hunk 2 above)
Comment (cooperp) [inline cmake/modules/AddSwift.cmake]: Actually it didn't move.  It the diff not really knowing how I applied the changes.

What we have before was a loop to set swiftlib_module_dependency_targets, then later a loop to set swiftlib_framework_depends_flattened, and finally a loop for swiftlib_private_link_libraries_targets.

In the original code, swiftlib_private_link_libraries_targets was initialized to swiftlib_module_dependency_targets, but we don't want that any more. This is because swiftlib_module_dependency_targets now contains references to the swift module files, and we want swiftlib_private_link_libraries_targets to reference the linked images.  So I changed swiftlib_private_link_libraries_targets to initialise to empty instead of swiftlib_module_dependency_targets then added appends to this loop to set it appropriately.

And yes, I could definitely move the in a separate commit then only make the change the following a change in a new commit:

list(APPEND swiftlib_module_dependency_targets
              "swift${mod}${VARIANT_SUFFIX}")

to

list(APPEND swiftlib_module_dependency_targets
              "swift${mod}${MODULE_VARIANT_SUFFIX}")

File: cmake/modules/SwiftSource.cmake (hunk 3 above)
Comment (cooperp) [inline cmake/modules/SwiftSource.cmake]: Oddly it didn't work with only -emit-module and no -o.  No error or anything, just a lack of anything being written to disk.  If I can remove the -emit-module-path then that would be fine for now.  I should file a bug for the lack of -o causing there to be no writes to disk and no errors.

File: cmake/modules/SwiftSource.cmake (hunk 5 above)
Comment (cooperp) [inline cmake/modules/SwiftSource.cmake]: if (apinotes_outputs) is much better.  I'll use that.  I just don't know enough CMake to know if thats ok or not.  I figured it would be on a string but wasn't sure about a list.

File: cmake/modules/SwiftSource.cmake (hunk 4 above)
Comment (cooperp) [inline cmake/modules/SwiftSource.cmake]: Yep, thats a mistake.  Good catch.  I might just always have this target run, but in some cases it will only create directories and in others it will create directories and api notes.  If you have a preference for something else i'm also totally fine with alternatives.

File: cmake/modules/SwiftSource.cmake (hunk 6 above)
Comment (cooperp) [inline cmake/modules/SwiftSource.cmake]: Good catch.  Will fix that.

File: cmake/modules/SwiftSource.cmake (hunk 7 above)
Comment (cooperp) [inline cmake/modules/SwiftSource.cmake]: Yep, I should put in "Generating {module_name}" instead.

File: cmake/modules/AddSwift.cmake (hunk 8 above)
Comment (cooperp) [inline cmake/modules/AddSwift.cmake]: Sure, will add documentation.

I can try out adding a parse argument.  Will need to also change 'target' above as I only added 'module_target' in this way to be consistent.

Would you prefer that change first, then land the rest of this?

File: cmake/modules/SwiftSource.cmake (hunk 4 above)
Comment (jrose-apple) [inline cmake/modules/SwiftSource.cmake]: That seems good enough. Creating API notes is very cheap.

File: cmake/modules/AddSwift.cmake (hunk 2 above)
Comment (jrose-apple) [inline cmake/modules/AddSwift.cmake]: There definitely is a move here: the ICU section and the swiftlib_framework_depends_flattened switched. I do see that it makes a little more sense.

File: cmake/modules/SwiftSource.cmake (hunk 12 above)
Comment (cooperp) [inline cmake/modules/SwiftSource.cmake]: It was actually just a move as the previous code had use adding '-emit-module ... -parse-as-library' on one branch and '-parse-as-library' on the other.  But I agree that this is getting a little unwieldy at its current size.  Thats probably worth someone refactoring a whole bunch of stuff though, which is better left to who is going to own it long term.

File: cmake/modules/AddSwift.cmake (hunk 2 above)
Comment (cooperp) [inline cmake/modules/AddSwift.cmake]: Sorry, yes there's a move.  I still had the 'git diff' my local machine shows in my head.  In it, it appears that the loop over swiftlib_framework_depends_flattened moves, which it didn't really.  What actually moved was the content I described above.

File: cmake/modules/AddSwift.cmake (hunk 10 above)
Comment (jrose-apple) [inline cmake/modules/AddSwift.cmake]: Right, it's still a dependency of the *link* target for each module.

File: cmake/modules/AddSwift.cmake [hunk 13]
```diff
@@ -1618,12 +1622,12 @@ function(add_swift_library name)
 
     _add_swift_library_single(
       ${name}
-      ""
       ${name}
       ${SWIFTLIB_SHARED_keyword}
       ${SWIFTLIB_STATIC_keyword}
       ${SWIFTLIB_OBJECT_LIBRARY_keyword}
       ${SWIFTLIB_SOURCES}
+      MODULE_TARGET ""
```
Comment (gottesmm) [inline cmake/modules/AddSwift.cmake]: This should be no longer necessary.

Comment (cooperp) [inline cmake/modules/AddSwift.cmake]: Thanks.  You're right.  Just removed it.

File: cmake/modules/SwiftSource.cmake (hunk 12 above)
Comment (gottesmm) [inline cmake/modules/SwiftSource.cmake]: Ok. I am fine with that.

**Comment by gottesmm** [comment]
I think you have answered all of my concerns. Removing my request.

**Review by gottesmm — APPROVED** [review]

**Comment by gottesmm** [comment]
Before you commit though, can you do a rebase/squash to prettify the history on your branch?

**Comment by cooperp** [comment]
Thanks Michael!  Yeah, will squash the commits and make it all pretty.  

Should I do one last swift-ci just to verify, or does that get done automatically for me when the PR is merged?

**Comment by cooperp** [comment]
@swift-ci Please test and merge
